### PR TITLE
osutil: allow using many globs in EnsureDirState

### PR DESCRIPTION
**Pull request by zyga**
This patch simply extends the directory synchronizer to allow more than
one glob to be used. This will come in handy when we want to provide
more than one kind of content for a given snap and then the
prefix/suffix (that is expressed as a glob) cannot capture that if only
one glob is available.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: osutil/syncdir.go [hunk 1]
```diff
@@ -63,16 +63,28 @@ var ErrSameState = fmt.Errorf("file state has not changed")
 // exhausted.
 //
 // In all cases, the function returns the first error it has encountered.
-func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, removed []string, err error) {
-	if _, err := filepath.Match(glob, "foo"); err != nil {
-		panic(fmt.Sprintf("EnsureDirState got invalid pattern %q: %s", glob, err))
+func EnsureDirStateMany(dir string, globs []string, content map[string]*FileState) (changed, removed []string, err error) {
+	for _, glob := range globs {
+		if _, err := filepath.Match(glob, "foo"); err != nil {
```
Comment (stolowski) [inline osutil/syncdir.go]: Oh, what is this? I realize it was in the existing code, but it looks perplexing.

Comment (chipaca) [reply osutil/syncdir.go]: maybe this'll help: `filepath.Glob("expr")` returns the list of files in the directory for which `filepath.Match` is true

Comment (chipaca) [reply osutil/syncdir.go]: Ah, now that I understand your question, the answer is that it's checking the globs syntax before actually doing any work

Comment (stolowski) [reply osutil/syncdir.go]: Ok I see, makes sense, thanks!

Comment (niemeyer) [reply osutil/syncdir.go]: Worth a comment in the code.

File: osutil/syncdir.go [hunk 2]
```diff
@@ -63,16 +63,28 @@ var ErrSameState = fmt.Errorf("file state has not changed")
...
+			panic(fmt.Sprintf("EnsureDirState got invalid pattern %q: %s", glob, err))
```
Comment (niemeyer) [inline osutil/syncdir.go]: Why are these using panics? Yes, it's bad, but it's much worse to have a non-functioning system altogether, that can't even revert. Feels like a bad decision we did.

Let's please also prefix the message with "internal error: ".

File: osutil/syncdir.go [hunk 3]
```diff
@@ -63,16 +63,28 @@ var ErrSameState = fmt.Errorf("file state has not changed")
 // exhausted.
 //
 // In all cases, the function returns the first error it has encountered.
-func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, removed []string, err error) {
-	if _, err := filepath.Match(glob, "foo"); err != nil {
-		panic(fmt.Sprintf("EnsureDirState got invalid pattern %q: %s", glob, err))
+func EnsureDirStateMany(dir string, globs []string, content map[string]*FileState) (changed, removed []string, err error) {
+	for _, glob := range globs {
```
Comment (niemeyer) [inline osutil/syncdir.go]: // Check syntax before doing anything.

File: osutil/syncdir.go [hunk 4]
```diff
@@ -63,16 +63,28 @@ var ErrSameState = fmt.Errorf("file state has not changed")
...
+		sane := false
+		for _, glob := range globs {
+			if ok, _ := filepath.Match(glob, baseName); ok {
+				sane = true
+				break
+			}
+		}
+		if !sane {
+			if len(globs) == 1 {
+				panic(fmt.Sprintf("EnsureDirState got filename %q which doesn't match the glob pattern %q", baseName, globs[0]))
```
Comment (niemeyer) [inline osutil/syncdir.go]: Same as above for the panic and prefix, and also no need to special case the 1 case for an internal error. The general message looks fine.

File: osutil/syncdir.go [hunk 5]
```diff
@@ -96,10 +108,14 @@ func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, r
 		changed = append(changed, baseName)
 	}
 	// Delete phase (remove files matching the glob that are not in content)
-	matches, err := filepath.Glob(filepath.Join(dir, glob))
-	if err != nil {
-		sort.Strings(changed)
-		return changed, nil, err
+	var matches []string
+	for _, glob := range globs {
```
Comment (niemeyer) [inline osutil/syncdir.go]: That's being done several times. If we move this logic out into a function the original code pretty much doesn't change:
```
func globAny(path, globs []string) (ok bool, error err)
```

File: osutil/syncdir.go [hunk 6]
```diff
@@ -120,6 +136,13 @@ func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, r
 	return changed, removed, firstErr
 }
 
+// EnsureDirState ensures that directory content matches expectations.
+//
+// This is like EnsureDirStateMany but it only supports one glob at a time.
+func EnsureDirState(dir string, glob string, content map[string]*FileState) (changed, removed []string, err error) {
+	return EnsureDirStateMany(dir, []string{glob}, content)
```
Comment (niemeyer) [inline osutil/syncdir.go]: s/Many/Globs/; It's not many dir, or many states, so the name seems misleading.

File: osutil/syncdir.go (hunk 2 above)
Comment (zyga) [reply osutil/syncdir.go]: I think you requested those back in the days but I cannot remember for sure.

**Pull request merged**